### PR TITLE
Cli testing

### PR DESCRIPTION
**Pull request by jsilverio22**
Issue: #868

### CODE REVIEW

**Comment by jsilverio22** [comment]
I branched this off of my acorn rm branch to write tests against that so this includes both unit testing for cli and acorn rm refactor changes @cjellick

File: pkg/cli/credential_logout.go [hunk 1]
```diff
@@ -8,8 +8,8 @@ import (
 	"github.com/spf13/cobra"
 )
 
-func NewCredentialLogout(root bool) *cobra.Command {
-	c := cli.Command(&CredentialLogout{}, cobra.Command{
+func NewCredentialLogout(root bool, cc client.CommandContext) *cobra.Command {
+	c := cli.Command(&CredentialLogout{client: cc.ClientFactory}, cobra.Command{
```
Comment (tylerslaton) [inline pkg/cli/credential_logout.go:12]: Same comment from `credential_login.go` applies here.

File: pkg/cli/run_test.go [hunk 2]
```diff
@@ -12,9 +13,9 @@ import (
 
 func ArgsToApp(t *testing.T, args ...string) *apiv1.App {
 	t.Helper()
-
+	client := &client.CommandContext{}
 	buf := &bytes.Buffer{}
-	cmd := NewRun(buf)
+	cmd := NewRun(buf, *client)
```
Comment (tylerslaton) [inline pkg/cli/run_test.go:18]: Any reason to create `client` as a pointer here? We just dereference it when we pass it to `NewRun()`

File: pkg/cli/all_test.go [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+				Output: "",
+			},
+			commandContext: client.CommandContext{
+				ClientFactory: &testdata.MockClientFactory{},
+				StdOut:        w,
+				StdErr:        w,
+				StdIn:         strings.NewReader("y\n"),
+			},
+			args: args{
+				args:   []string{"-oyaml"},
```
Comment (tylerslaton) [inline pkg/cli/all_test.go:85]: What's the reasoning behind not doing this?
```suggestion
				args:   []string{"-o", "yaml"},
```

File: pkg/cli/log_test.go [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+				StdOut:        w,
+				StdErr:        w,
+				StdIn:         strings.NewReader(""),
+			},
+			args: args{
+				args:   []string{"found"},
+				client: &testdata.MockClient{},
+			},
+			wantErr: false,
+			wantOut: "",
```
Comment (tylerslaton) [inline pkg/cli/log_test.go:51]: This looks like the success state, should we not have some output here?

File: pkg/cli/render_test.go [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+	tests := []struct {
+		name           string
+		fields         fields
+		args           args
+		wantErr        bool
+		wantOut        string
+		commandContext client.CommandContext
+	}{
+		{
+			name: "acorn render .", fields: fields{
```
Comment (tylerslaton) [inline pkg/cli/render_test.go:35]: We should test the happy path for this as well.

**Review by tylerslaton — CHANGES_REQUESTED** [review]
Awesome work. I left some comments for my first pass. I'll come back through and mainly take a look at the `MockClient` since I didn't have time for it this morning.

File: pkg/cli/log_test.go (hunk 4 above)
Comment (jsilverio22) [reply pkg/cli/log_test.go:51]: there are not any logs being outputted by that container so no output

File: pkg/cli/acorn.go [hunk 6]
```diff
@@ -20,35 +21,41 @@ func New() *cobra.Command {
...
+		NewUninstall(cmdContext),
+		NewInfo(cmdContext),
+		NewLogs(cmdContext),
+		NewCredentialLogin(true, cmdContext),
+		NewCredentialLogout(true, cmdContext),
+		NewPull(cmdContext),
+		NewPush(cmdContext),
+		NewRm(cmdContext),
+		NewRun(os.Stdout, cmdContext),
+		NewUpdate(os.Stdout, cmdContext),
```
[the remark this answers is not in the document]
Comment (jsilverio22) [reply pkg/cli/acorn.go]: removed

File: pkg/cli/acorn_test.go [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+	tests := []struct {
+		name           string
+		fields         fields
+		args           args
+		wantErr        bool
+		wantOut        string
+		commandContext client.CommandContext
+	}{
+		{
+			name: "acorn info", fields: fields{
```
[the remark this answers is not in the document]
Comment (jsilverio22) [reply pkg/cli/acorn_test.go]: reformatted this

File: pkg/cli/acorn_test.go [hunk 8]
```diff
@@ -0,0 +1,70 @@
...
+		},
+	}
+	for _, tt := range tests {
+		r, w, _ := os.Pipe()
+		os.Stdout = w
+		tt.args.cmd = New()
+		tt.args.cmd.SetArgs(tt.args.args)
+		err := tt.args.cmd.Execute()
+		if err != nil && !tt.wantErr {
+			assert.Failf(t, "got err when err not expected", "got err")
```
[the remark this answers is not in the document]
Comment (jsilverio22) [reply pkg/cli/acorn_test.go]: added to all tests

File: pkg/cli/credential_login.go [hunk 9]
```diff
@@ -13,8 +13,8 @@ import (
 	apierror "k8s.io/apimachinery/pkg/api/errors"
 )
 
-func NewCredentialLogin(root bool) *cobra.Command {
-	c := cli.Command(&CredentialLogin{}, cobra.Command{
+func NewCredentialLogin(root bool, cc client.CommandContext) *cobra.Command {
+	c := cli.Command(&CredentialLogin{client: cc.ClientFactory}, cobra.Command{
```
[the remark this answers is not in the document]
Comment (jsilverio22) [reply pkg/cli/credential_login.go]: refactored for consistency

File: pkg/client/client.go [hunk 10]
```diff
@@ -13,10 +11,24 @@ import (
...
 
-func Default() (Client, error) {
+type CommandContext struct {
+	ClientFactory ClientFactory
+	StdOut        *os.File
+	StdErr        *os.File
+	StdIn         *strings.Reader
+}
+type ClientFactory interface {
+	CreateDefaultClient() (Client, error)
```
[the remark this answers is not in the document]
Comment (jsilverio22) [reply pkg/client/client.go]: refactored name

File: pkg/cli/run_test.go (hunk 2 above)
Comment (jsilverio22) [reply pkg/cli/run_test.go]: no reason I thought it was used differently in the NewRun but changed from pointer

**Review by cjellick — APPROVED** [review]
just need the conflict resolved

**Review by cjellick — APPROVED** [review]

**Review by tylerslaton — APPROVED** [review]
Awesome work @jsilverio22

**Pull request merged**